### PR TITLE
exo-generate

### LINKED ISSUE #749: new command: exo generate
Generate all the docker-compose and terraform files. 

`run`, `test` and `deploy` all do this at the beginning of there command.

Pair with `exo needs-generate`which checks if the docker-compose and terraform files are up to date.

@hugobho what do you think? @alexdavid and i were discussing this.

### PR DESCRIPTION
**Pull request by hugoesthere**
partially resolves https://github.com/Originate/exosphere/issues/749

Not sure what the best way to test this is, but seeing as the crux of this command is `GetApplicationDockerConfigs`, I think we have most of our test cases covered.

### CODE REVIEW

**Comment by charlierudolph** [comment]
This should also generate the terraform file. Lets also add a feature test that verifies the files are added (without testing the content)

**Comment by hugoesthere** [comment]
@charlierudolph comments addressed

**Comment by charlierudolph** [comment]
Should the base branch be updated to `hh-terraform-vars` since it requires #775?

**Comment by hugoesthere** [comment]
Yes, on my machine that's the case I just didn't update it here

File: bin/pre-release [hunk 1]
```diff
@@ -3,6 +3,6 @@
 hash=$(git log --format="%h" -n 1 terraform)
 
 if [ -z "$(grep "$hash" src/cmd/deploy.go)" ]; then
```
Comment (charlierudolph) [inline bin/pre-release]: this also needs to be updated to shared

File: features/generate.feature [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  Scenario: generate docker-compose and terraform files
+    Given I am in the root directory of the "simple" example application
+    When starting "exo generate" in my application directory
+    Then it prints "docker-compose and terraform files generated" in the terminal
+    And my workspace contains the files:
+      | docker-compose/run_development.yml |
+      | docker-compose/run_production.yml  |
+      | docker-compose/test.yml            |
+      | terraform/main.tf                  |
+
```
Comment (charlierudolph) [inline features/generate.feature]: Please remove this extra empty line

File: src/cmd/generate.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+		context, err := GetContext()
+		if err != nil {
+			log.Fatal(err)
+		}
+		err = application.GenerateComposeFiles(context.AppContext)
+		if err != nil {
+			log.Fatal(err)
+		}
+		deployConfig := getBaseDeployConfig(context.AppContext)
+		writer := os.Stdout
```
Comment (charlierudolph) [inline src/cmd/generate.go]: is this variable needed? Its only used on the line below

**Review by charlierudolph — APPROVED** [review]

File: src/cmd/generate.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		err = application.GenerateComposeFiles(context.AppContext)
+		if err != nil {
+			log.Fatal(err)
+		}
+		deployConfig := getBaseDeployConfig(context.AppContext)
+		err = application.GenerateTerraformFiles(deployConfig)
+		if err != nil {
+			log.Fatal(err)
+		}
+		fmt.Println("docker-compose and terraform files generated")
```
Comment (charlierudolph) [inline src/cmd/generate.go]: I don't think we need either of these messages (generating or generated)

Comment (hugoesthere) [reply src/cmd/generate.go]: We don't, I really only have them there so the tests have some text to wait on

Comment (charlierudolph) [reply src/cmd/generate.go]: We can wait for the process to exit.

**Pull request merged**